File: examples/12_redos/vulnerable.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

const server = new McpServer({ name: "redos-demo", version: "1.0.0" });

// VULNERABLE: Catastrophic backtracking regex applied to user input (CWE-1333)
const EMAIL_RE = /^([a-zA-Z0-9]+)*@[a-zA-Z0-9]+\.[a-zA-Z]+$/;  // nested quantifier: (x+)*
const SLUG_RE  = /(a+)+$/;                                         // classic ReDoS pattern

server.tool("validate_email", { email: z.string() }, async ({ email }) => {
  // Attacker sends "aaaaaaaaaaaaaaaaaaa!" — causes exponential backtracking
  const valid = EMAIL_RE.test(email);
  return { content: [{ type: "text", text: valid ? "Valid" : "Invalid" }] };
});

server.tool("validate_slug", { slug: z.string() }, async ({ slug }) => {
  // Even worse: /(a+)+$/ — O(2^n) for strings like "aaaaaaaaaaaaaaaaaaaX"
  const match = SLUG_RE.test(slug);
  return { content: [{ type: "text", text: match ? "Match" : "No match" }] };
});

File: examples/12_redos/vulnerable.py
"""
Vulnerable: ReDoS — catastrophic regex backtracking — REDOS

Run: mcp-sentinel scan examples/12_redos/vulnerable.py
Expected: HIGH REDOS findings
"""
import re
from mcp.server.fastmcp import FastMCP

mcp = FastMCP("Vulnerable Server")

@mcp.tool()
def validate_input(user_input: str) -> bool:
    """Validate user input format — VULNERABLE to ReDoS."""
    # Attack: user_input = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"
    # This causes exponential backtracking — locks the thread for seconds/minutes
    pattern = re.compile(r"(a+)+$")
    return bool(pattern.match(user_input))

@mcp.tool()
def validate_email(email: str) -> bool:
    """Validate email format — VULNERABLE to ReDoS."""
    # Nested quantifier: (\w+\s*)+ causes catastrophic backtracking
    return bool(re.match(r"^(\w+\s*)+@[\w.]+$", email))

@mcp.tool()
def check_log_format(log_line: str) -> bool:
    """Validate log line format — VULNERABLE to ReDoS."""
    # ([\w\-]+)+ with no end-anchor causes exponential blowup
    pattern = r"^([\w\-]+)+ (ERROR|WARN|INFO)"
    return bool(re.match(pattern, log_line))
